### ISSUE TITLE
[Functions]Templates for RabbitMQ

### ISSUE THREAD

**Issue by yojagad** [microsoft/azure-maven-plugins#1197]
In effort to GA RabbitMQ extension, we currently have templates for non-dotnet languages
https://github.com/Azure/azure-functions-templates/pull/995

We need templates for java as well.

Related java library that defines annotations can be found in the repository. https://www.github.com/Azure/azure-functions-rabbitmq-extension

cc: @cachai2

**Comment by pragnagopa** [comment]
@Flanker32 / @andxu - please let us know if you need any info from us.

**Comment by Flanker32** [comment]
@pragnagopa So we will have an new trigger type `RabbitMQTrigger` and need to support it in maven tooling right? Will it be covered in `azure-functions-java-library` or it will keep in `azure-functions-rabbitmq-extension` and will be treated as a CustomBinding?

@yojagad For the templates, do you mean we need to provide a sample project for this trigger? For now we don't have java function templates in [Azure/azure-functions-templates](https://github.com/Azure/azure-functions-templates) but have sample project in [AzureSample](https://github.com/Azure-Samples/azure-functions-samples-java)

**Comment by yojagad** [comment]
@Flanker32 This is part of `azure-functions-rabbitmq-extension` treated as a custom binding (https://github.com/Azure/azure-functions-rabbitmq-extension/tree/dev/binding-library/java) . Version 1.0.0-beta of `azure-functions-java-library-rabbit` was released on mvn central a week or so ago. Templates specified in https://github.com/Azure/azure-functions-templates allow quick creation of functions for command line (via `func init` and `func new` commands), vscode and visual studio as well for non-java languages.

We're looking for support that would enable similar creation experience as part of command line (maybe something like `mvn azure-functions:add -Dfunctions.package=com.your.package -Dfunctions.name=NewFunction -Dfunctions.template=RabbitMQTrigger
`), vscode, eclipse, intellij, etc for java. The resulting, generated java function could look something like below

````Java

package com.function;

import com.microsoft.azure.functions.ExecutionContext;
import com.microsoft.azure.functions.annotation.FunctionName;
import com.microsoft.azure.functions.rabbitmq.annotation.RabbitMQTrigger;

/**
 * Azure Functions with RabbitMQ Trigger.
 */
public class Function {
    @FunctionName("RabbitMQExample")
    public void run(
            @RabbitMQTrigger(
                connectionStringSetting = "ConnectionString",
                queueName = "TestQueue"
            ) String input,
            final ExecutionContext context) {
        context.getLogger().info("Java HTTP trigger processed a request." + input);
    }
}

````

**Comment by pragnagopa** [comment]
@Flanker32 - @yojagad added details. Yes, We are looking for similar experience to other triggers / output bindings supported by mvn plugin

**Comment by Flanker32** [comment]
@yojagad @pragnagopa, Thank you all for your updates and sorry for the late reply, so here is the work items for the new trigger:
- Verify maven plugin could work with the new annotation, as `RabbitMQ` is a CustomBinding which should be supported in current release
- Add template in `azure-functions:add` for `RabbitMQ`, which is also used in VSCode Azure Function Extensions
    * It seems that there are more parameters in [`RabbitMQTrigger`](https://github.com/Azure/azure-functions-rabbitmq-extension/blob/dev/binding-library/java/src/main/java/com/microsoft/azure/functions/rabbitmq/annotation/RabbitMQTrigger.java), should we ask users to set values for all of them or only `connectionStringSetting` and `queueName` is enough?
    * As `RabbitMQ` is a CustomBinding which is not included in azure-functions-java-library, do we also need to add the dependency in pom.xml, or prompt a hint message, or the options in only valid if `azure-functions-rabbitmq-extension` exists in users' dependencies?

**Comment by yojagad** [comment]
``` quote
It seems that there are more parameters in RabbitMQTrigger, should we ask users to set values for all of them or only connectionStringSetting and queueName is enough?
```

Since we're only looking to create a QuickStart experience, the fastest way to start would be to prompt just for `connectionStringSetting` and `queueName` . The binding works in a way such that if `connectionStringSetting` is set, it ignores the rest of the parameters even if present.



 ``` quote
As RabbitMQ is a CustomBinding which is not included in azure-functions-java-library, do we also need to add the dependency in pom.xml, or prompt a hint message, or the options in only valid if azure-functions-rabbitmq-extension exists in users' dependencies?
```

Yes, it would be great if the `azure-functions-java-library-rabbitmq` is automatically included as an additional dependency in pom.xml in addition to `azure-functions-java-library`

**Comment by cachai2** [comment]
Is there an ETA for when we could get MVN templates for RabbitMQ?

**Comment by Flanker32** [comment]
@pragnagopa @yojagad @cachai2 Really sorry for the late response, we will add this feature in our next sprint release(end of April). There might be some more tasks to add the dependency, so for the first version, we will just add the template first, and show warning for users to add the dependency. 
Do you have any link for doc so that we could show in warning to help user add the dependency? Or we can add the steps in the template as comment as user may follow the instruction. Which do you prefer?

**Comment by pragnagopa** [comment]
@Flanker32  - thanks for getting back to us!

@cachai2 - please chime on preferred user experience

@yojagad - is there azure-functions-java-library-rabbitmq 1.0.0 package - without beta flag?

**Comment by cachai2** [comment]
@Flanker32 Thank you for the update. It would be great if we could add the steps in the template as a comment to reduce the need for customers to bounce around from the template to docs to get the necessary info.

@yojagad I remember that Fabio OK'd the package with beta flag for GA, but it would be good to update the package to remove the beta flag as it is creating some confusion.

**Comment by Flanker32** [comment]
@cachai2 @pragnagopa It seems that there is only 1.0.0-beta version in maven repository, so shall we add the link to sonatype  in comment and users could always get the latest version there? Or do we have any docs could be used here?


``` java
package com.function;

import com.microsoft.azure.functions.ExecutionContext;
import com.microsoft.azure.functions.annotation.FunctionName;
import com.microsoft.azure.functions.rabbitmq.annotation.RabbitMQTrigger;

/**
 * Azure Functions with RabbitMQ Trigger.
 *
 * Please add com.microsoft.azure.functions:azure-functions-java-library-rabbitmq to your project dependencies
 * You may get the latest version in https://search.maven.org/artifact/com.microsoft.azure.functions/azure-functions-java-library-rabbitmq
 * 
 */
public class Function {
    @FunctionName("RabbitMQExample")
    public void run(
            @RabbitMQTrigger(
                    connectionStringSetting = "ConnectionString",
                    queueName = "TestQueue"
            ) String input,
            final ExecutionContext context) {
        context.getLogger().info("Java HTTP trigger processed a request." + input);
    }
}
```

**Comment by pragnagopa** [comment]
@Flanker32 - Thanks for getting back to us! We now have GA package without beta flag. @yojagad - please publish to mvn

**Comment by pragnagopa** [comment]
Tracking issue: https://github.com/Azure/azure-functions-rabbitmq-extension/issues/160 for publishing to mvn

**Comment by yojagad** [comment]
@pragnagopa @Flanker32 @cachai2 1.0.0 is now up on mvn https://mvnrepository.com/artifact/com.microsoft.azure.functions/azure-functions-java-library-rabbitmq

**Comment by Flanker32** [comment]
@yojagad Thanks for your updates! We will add the [template](https://github.com/microsoft/azure-maven-plugins/issues/1197#issuecomment-805588329) in our next release.

**Comment by Flanker32** [comment]
Pr for this item : https://github.com/microsoft/azure-maven-plugins/pull/1393

**Comment by Flanker32** [comment]
RabbitMQ support has been released with [functions maven plugin 1.11.0](https://search.maven.org/artifact/com.microsoft.azure/azure-functions-maven-plugin/1.11.0/maven-plugin)
